### ISSUE TITLE
`sig cat` selection relies on md5sum, which can pull in unwanted sigs with duplicated md5

### ISSUE THREAD

**Issue by bluegenes** [sourmash-bio/sourmash#2617]
tl;dr - **`sourmash sig cat` uses `md5` to select signatures, and this creates an issue when trying to build a database by selecting signatures from wort. This means that trying to generate `gtdb-rs214` with 402k signatures yields a zipfile with 687k signatures (includes closely-related strains with identical `md5sums`).**

> work on farm: `/home/ntpierce/database-releases/gtdb-rs214`

## db building workflow
To build a `gtdb-rs214` database release, I adapted the genbank database procedure.
> Snakefile on github: https://github.com/sourmash-bio/database-releases/blob/gtdb-rs214/gtdb-rs214.genomic/Snakefile

1. I made updated the the `wort` signature manifest using the update code: /group/ctbrowngrp/sourmash-db/wort-manifests/ to create `2023-05-05.wort.sqlmf` (sqlite manifest).

2. I used the wort manifest with the gtdb-lineages file to build a `gtdb-rs214` manifest.

```
 sourmash sig check \
             --picklist gtdb-rs214.lineages.csv:ident:ident \
            2023-05-05.wort.sqlmf \
            --output-missing gtdb-rs214.manifest.csv \
            --save-manifest gtdb-rs214.manifest.csv
```
> Note this took a few days, but we had all the signatures!

3. I built a `gtdb-rs214` zipfile database using this manifest and sig cat, like so:

```
sourmash sig cat gtdb-rs214.manifest.csv -k 21 -o gtdb-rs214-k21.abund.bak.zip 
```

>(`bak` added during troubleshooting, writing here for consistency)

## Issue: Zipfile has more signatures than would be expected

Let's look at the manifest and lineages:

lineages file has 402k unique idents:
```
$wc -l gtdb-rs214.lineages.csv

402710
```
So 402709 lines without the header
> ```
> ident,gtdb_representative,superkingdom,phylum,class,order,family,genus,species
>GCA_000008085.1,t,d__Archaea,p__Nanoarchaeota,c__Nanoarchaeia,o__Nanoarchaeales,f__Nanoarchaeaceae,g__Nanoarchaeum,s__Nanoarchaeum equitans
>GCA_000016605.1,t,d__Archaea,p__Thermoproteota,c__Thermoprotei_A,o__Sulfolobales,f__Sulfolobaceae,g__Metallosphaera,s__Metallosphaera sedula
> ```

The manifest matches this lineage
```
less gtdb-rs214.manifest.csv | grep "21,DNA" | wc -l
402709
less gtdb-rs214.manifest.csv | grep "31,DNA" | wc -l
402709
less gtdb-rs214.manifest.csv | grep "51,DNA" | wc -l
420526
```
> note: be sure to grep with `21,DNA` as `21,` might be found in the md5 column :)

> ```
> # SOURMASH-MANIFEST-VERSION: 1.0
>internal_location,md5,md5short,ksize,moltype,num,scaled,n_hashes,with_abundance,name,filename
>/group/ctbrowngrp/sourmash-db/wort-manifests/2023-05-05.wort.sqlmf,064d8881974f960c788f5e9a2eb5874f,064d8881,21,DNA,0,1000,2911,0,"GCA_025056565.1 Chloroherpetonaceae bacterium, ASM2505656v1",/dev/fd/63
>/group/ctbrowngrp/sourmash-db/wort-manifests/2023-05-05.wort.sqlmf,dec43c32291fcbcdb07dc7775a7d72bb,dec43c32,31,DNA,0,1000,3029,0,"GCA_025056565.1 Chloroherpetonaceae bacterium, ASM2505656v1",/dev/fd/63
>/group/ctbrowngrp/sourmash-db/wort-manifests/2023-05-05.wort.sqlmf,de5c11f84e880bcbe64d955dfe96cd2f,de5c11f8,51,DNA,0,1000,2859,0,"GCA_025056565.1 Chloroherpetonaceae bacterium, ASM2505656v1",/dev/fd/63
>/group/ctbrowngrp/sourmash-db/wort-manifests/2023-05-05.wort.sqlmf,9d78fdcc950594fbfd46dbe548f970be,9d78fdcc,21,DNA,0,1000,2598,0,"GCA_023393635.1 Proteobacteria bacterium, ASM2339363v1",/dev/fd/63
>/group/ctbrowngrp/sourmash-db/wort-manifests/2023-05-05.wort.sqlmf,4278a25f3b743edddf446cd4a1e736de,4278a25f,31,DNA,0,1000,2642,0,"GCA_023393635.1 Proteobacteria bacterium, ASM2339363v1",/dev/fd/63
>/group/ctbrowngrp/sourmash-db/wort-manifests/2023-05-05.wort.sqlmf,941fdbdfd0a57919fa4bd648378b79a4,941fdbdf,51,DNA,0,1000,2604,0,"GCA_023393635.1 Proteobacteria bacterium, ASM2339363v1",/dev/fd/63
> ```

This manifest does contain some duplicated md5sums (id'd by `sig cat`):

```
sourmash sig cat gtdb-rs214-k21.abund.bak.zip --picklist gtdb-rs214.manifest.csv::manifest --ksize 21 -o gtdb-rs214-k21.abund.zip

== This is sourmash version 4.8.3.dev0. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

picking column 'md5' of type 'md5' from 'gtdb-rs214.manifest.csv'
loaded 1176290 distinct values into picklist.
WARNING: 20055 values in picklist column 'md5' were not distinct
```
> note I am currently using `try-submanifest` branch in case I need the `sig grep` code I implemented, but it doesn't have any differences from v4.8.2 in `sig cat`

### ...But this doesn't explain all extra signatures 

Given the duplicated md5sums, above, I would expect 20k extra signatures in the database. Instead, there are ~285k extra!??

```
(sourmash_dev) ntpierce@farm:~/database-releases/gtdb-rs214.genomic$ sourmash sig summarize gtdb-rs214-k21.abund.bak.zip

== This is sourmash version 4.8.3.dev0. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

** loading from 'gtdb-rs214-k21.abund.bak.zip'
path filetype: ZipFileLinearIndex
location: /home/ntpierce/database-releases/gtdb-rs214.genomic/gtdb-rs214-k21.abund.bak.zip
is database? yes
has manifest? yes
num signatures: 687192
** examining manifest...
total hashes: 2667245979
summary of sketches:
   687192 sketches with DNA, k=21, scaled=1000, abund 2667245979 total hashes
```

Let's output zipfile manifest csv + count names, md5s

```
 sourmash sig manifest gtdb-rs214-k21.abund.bak.zip --csv gtdb-rs214-k21.abund.bak.zip.manifest.csv --no-rebuild-manifest
```

count unique values in `name` column:
```
(sourmash_dev) ntpierce@bm2:~/database-releases/gtdb-rs214.genomic$ less gtdb-rs214-k21.abund.bak.zip.manifest.csv | cut -f 10  -d ',' | sort | uniq | wc -l
687194
```
count unique values in `md5` column:
```
(sourmash_dev) ntpierce@bm2:~/database-releases/gtdb-rs214.genomic$ less gtdb-rs214-k21.abund.bak.zip.manifest.csv
(sourmash_dev) ntpierce@bm2:~/database-releases/gtdb-rs214.genomic$ less gtdb-rs214-k21.abund.bak.zip.manifest.csv | cut -f 2  -d ',' | sort | uniq | wc -l
389386
```
... so yes, looks like it's still an `md5sum` issue

### So what are the extra signatures?

Checking the numbers,
402709 in `gtdb-rs114.lineages.csv`
402709 in `gtdb-rs214` manifest created with `sig check`
- 20055 were duplicated.
687194 unique `name` values in zipfile manifest
389386 unique `md5` values in zipfile manifest

Likely we have some md5's duplicated many times (to get so many extra signatures). Let's check:
```
less gtdb-rs214-k21.abund.bak.zip.manifest.csv | cut -f 2 -d ',' |  sort | uniq -c | sort
```
tail end of output:
```
     94 ec004afb92635350f1e2888d80f0180e
    102 d1d7d5654074731677fcd995f6dbd122
    106 69c48502c07e6672d0b4bf10429b15fa
    128 a304efc55ab87be8eef6a5d9e949bd61
    132 784259ecf3329bd13f854e22ff8896c2
    149 3500248b7dd72beecb1e2637dade710d
    154 2cd58b27239a1f9c5ff63a1af3cccd92
    284 cbcc7a8cb59ba76820545a1e33484875
    366 e79a980fa9061a5906a7c42c03edb174
    626 6e4b9859a06a30aa6a7db3a4ddfeba14
```

Looking at that last one (with 626):
```
 grep 6e4b9859a06a30aa6a7db3a4ddfeba14 gtdb-rs214-k21.abund.bak.zip.manifest.csv | cut -f 10 -d ',' | sort | less
```
> add grep -v "pertussis" to check

All 626 are strains of `Bordetella pertussis`!

Second to last (with 366), same commands -- all are strains of `Francisella tularensis`.

**Ok so the extra signatures are closely-related strains with identical `md5sums`.**

### How do we solve?

1. simple workaround using https://github.com/sourmash-bio/sourmash/pull/2602:

Use `sig extract` with the database and original manifest file:
```
sourmash sig extract gtdb-rs214-k21.abund.bak.zip -o gtdb-rs214-k21.abund.zip --picklist gtdb-rs214.manifest.csv:name:name --ksize 21
```
> Note that  --picklist gtdb-rs214.manifest.csv::manifest does _not_ work (selects all)

2. long-term solution:

Can we add manifest-style checking to `sig cat` (similar to https://github.com/sourmash-bio/sourmash/pull/2602) so that after getting the matching `md5sum` signatures, we check e.g. name/ident, etc and only keep the exact signatures we want? We may also want to enable manifest checking more generally, given that the above solution fails (continues to be incorrect) if you pass `--picklist gtdb-rs214.manifest.csv::manifest`

**Comment by ctb** [comment]
#2747 adds an API-level demonstration of this problem.
